File: observability/logger.js
const fs = require('fs');
const path = require('path');

// Ensure log directory exists at the root
const logDir = path.join(__dirname, '../logs');
if (!fs.existsSync(logDir)) {
  fs.mkdirSync(logDir, { recursive: true });
}

const formatLog = (level, message, service = 'backend', metadata = {}) => {
  return JSON.stringify({
    timestamp: new Date().toISOString(),
    level: level.toUpperCase(),
    service,
    message,
    ...metadata
  });
};

const logger = {
  info: (message, service = 'backend', metadata = {}) => {
    const log = formatLog('INFO', message, service, metadata);
    console.log(log);
    try {
      fs.appendFileSync(path.join(logDir, 'combined.log'), log + '\n');
    } catch (err) {
      // Fail-silent on read-only filesystems
    }
  },
  error: (message, service = 'backend', metadata = {}) => {
    const log = formatLog('ERROR', message, service, metadata);
    console.error(log);
    try {
      fs.appendFileSync(path.join(logDir, 'error.log'), log + '\n');
      fs.appendFileSync(path.join(logDir, 'combined.log'), log + '\n');
    } catch (err) {
      // Fail-silent
    }
  },
  warn: (message, service = 'backend', metadata = {}) => {
    const log = formatLog('WARN', message, service, metadata);
    console.warn(log);
    try {
      fs.appendFileSync(path.join(logDir, 'combined.log'), log + '\n');
    } catch (err) {
      // Fail-silent
    }
  },
  debug: (message, service = 'backend', metadata = {}) => {
    if (process.env.NODE_ENV !== 'production') {
      const log = formatLog('DEBUG', message, service, metadata);
      console.log(log);
    }
  }
};

module.exports = logger;

File: observability/logger.py
import json
import os
import sys
from datetime import datetime

LOG_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), 'logs')
if not os.path.exists(LOG_DIR):
    try:
        os.makedirs(LOG_DIR, exist_ok=True)
    except Exception:
        pass

def format_log(level, message, service='agents', metadata=None):
    if metadata is None:
        metadata = {}
    log_data = {
        "timestamp": datetime.utcnow().isoformat() + "Z",
        "level": level.upper(),
        "service": service,
        "message": message
    }
    log_data.update(metadata)
    return json.dumps(log_data)

class Logger:
    @staticmethod
    def info(message, service='agents', metadata=None):
        log = format_log('INFO', message, service, metadata)
        print(log, file=sys.stdout)
        try:
            with open(os.path.join(LOG_DIR, 'combined.log'), 'a', encoding='utf-8') as f:
                f.write(log + '\n')
        except Exception:
            pass

    @staticmethod
    def error(message, service='agents', metadata=None):
        log = format_log('ERROR', message, service, metadata)
        print(log, file=sys.stderr)
        try:
            with open(os.path.join(LOG_DIR, 'error.log'), 'a', encoding='utf-8') as f:
                f.write(log + '\n')
            with open(os.path.join(LOG_DIR, 'combined.log'), 'a', encoding='utf-8') as f:
                f.write(log + '\n')
        except Exception:
            pass

    @staticmethod
    def warn(message, service='agents', metadata=None):
        log = format_log('WARN', message, service, metadata)
        print(log, file=sys.stdout)
        try:
            with open(os.path.join(LOG_DIR, 'combined.log'), 'a', encoding='utf-8') as f:
                f.write(log + '\n')
        except Exception:
            pass

    @staticmethod
    def debug(message, service='agents', metadata=None):
        if os.environ.get('NODE_ENV') != 'production' and os.environ.get('PYTHON_ENV') != 'production':
            log = format_log('DEBUG', message, service, metadata)
            print(log, file=sys.stdout)
